### PR TITLE
Update for latest Julia

### PR DESCRIPTION
**Pull request by smolck**
Not quite there yet, tests still don't pass and I doubt the code acts exactly the same as it did before, but it's a start.

### CODE REVIEW

File: src/client.jl [hunk 1]
```diff
@@ -27,24 +32,24 @@ end
 # this is probably not most efficient in the common case (no contention)
 # but it's the simplest way to assure task-safety of reading from the stream
 function readloop(c::NvimClient, handler)
-    while !eof(c.output)
-        msg = unpack(c.output)
-        kind = msg[1]::Int
+    while !eof(c.proc)
+        msg = unpack(c.proc.out)
```
Comment (smolck) [inline src/client.jl]: TODO: Not sure on the `proc` vs. `proc.out` here

**Comment by smolck** [comment]
Not really sure why, but right now if I run `$ julia -e 'using Pkg; Pkg.instantiate(); Pkg.develop(PackageSpec(path=pwd())); Pkg.test("Neovim")'` to run the tests, I get this strange error `ERROR: LoadError: UndefVarError: c not defined` at line 15 of runtests.jl. I think it has something to do with the generated api, but I'm not really sure why it happens. It happened after the changes to api_gen.jl in [this commit](https://github.com/bfredl/Neovim.jl/pull/11/commits/987669ef4c57b87d8a13b8ef1321ff5710dec6a5) I think . . .

EDIT: Fixed it, the error was from me deleting the code to properly generate the API functions with the `c` argument set to something.

File: src/plugin_host.jl [hunk 2]
```diff
@@ -120,7 +120,7 @@ macro fnsync(args...)
 end
 
 function fun(ex)
-    if ex.head == :block && length(ex.args) == 2 && ex.args[1].head == :line
+    if ex.head == :block && length(ex.args) == 2 && ex.args[2].head == :call
```
Comment (smolck) [inline src/plugin_host.jl]: TODO: This is pointless I think

**Comment by smolck** [comment]
@bfredl This branch is ready for review I think, it's mostly done (in fact the remote plugin things should work), the tests just aren't passing at the moment for some reason.

Note that in updating the code I may have accidentally changed a few things that I shouldn't have (although I'm not sure where if that's the case). I also need to go through the code removing some `println` statements and unnecessary comments and such that I used while working on this.

(To run the tests btw you can run `julia -e 'using Pkg; Pkg.instantiate(); Pkg.develop(PackageSpec(path=pwd())); Pkg.test("Neovim")'`)

File: .travis.yml [hunk 3]
```diff
@@ -1,11 +1,9 @@
 language: julia
 julia:
-  - 0.4
+  - 1.5.0
```
Comment (bfredl) [inline .travis.yml:4]: just curious, does this work on earlier versions (like 1.2) or uses it an 1.5 feature?

File: src/Neovim.jl [hunk 4]
```diff
@@ -9,38 +9,41 @@ export Buffer, Tabpage, Window
...
-    input, output = STDOUT, STDIN
+    input, output = stdout, stdin
     if haskey(ENV, "NEOVIM_JL_DEBUG")
         debug = open(ENV["NEOVIM_JL_DEBUG"], "w")
         redirect_stdout(debug)
         redirect_stderr(debug)
     else
-        redirect_stdout()
-        redirect_stderr()
+        # redirect_stdout()
```
Comment (bfredl) [inline src/Neovim.jl]: is this WIP?

File: .travis.yml (hunk 3 above)
Comment (smolck) [reply .travis.yml:4]: Good question, I’m not sure, I should find a way to test that . . . I imagine it’s at least compatible with a few previous versions, but I don’t know which ones exactly.

File: src/Neovim.jl (hunk 4 above)
Comment (smolck) [reply src/Neovim.jl]: Good catch, I think those can be safely uncommented now. I thought they were causing some issues but I think the cause was actually something else.

File: .travis.yml (hunk 3 above)
Comment (smolck) [reply .travis.yml:4]: @bfredl I looked at [Flux's](https://github.com/FluxML/Flux.jl/blob/cd811821a0ea150a72f8782b5ddf8b9bb5d1ef69/.travis.yml#L8) and [Gadfly's](https://github.com/GiovineItalia/Gadfly.jl/blob/ff151918958fb6d975d8e3af368d7eceecc877db/.travis.yml#L4) `.travis.yml`'s just to see what versions they were compatible with, and they both appear to be compatible with 1.0, so we could probably do the tests on `1` as well and see if that works. If not we can adjust them later.

**Comment by smolck** [comment]
@bfredl Could you give this another look over when you get the chance? If it all looks good I’ll merge it, maybe w/some doc updates if necessary. 

Note that some of the [tests about type stability](https://github.com/smolck/Neovim.jl/blob/f29c398ee6f17074d374c62e2813dc154105aff4/test/runtests.jl#L136) still don’t pass, but I don’t think that should block merge, would you agree?

File: src/repl.jl [hunk 5]
```diff
@@ -3,27 +3,34 @@
...
+mutable struct NvimReplState
     active::Bool
-    nv::NvimClient
+    nv::Neovim.NvimClient
     s::LineEdit.MIState
     rbuf::IOBuffer
     function NvimReplState()
         state = new(false)
-        state.nv, proc = nvim_spawn(state)
+        state.nv = Neovim.nvim_spawn(state)
```
Comment (smolck) [inline src/repl.jl:22]: ~~Should add that proc part back~~

EDIT: actually, don't remember why I said this . . .

File: src/repl.jl [hunk 6]
```diff
@@ -3,27 +3,34 @@
...
-import Neovim: get_cursor, set_cursor, command, input
+# import Base: LineEdit, REPL
+import REPL
+import REPL.LineEdit
 
-type NvimReplState
+include("Neovim.jl")
+
+# using Neovim
+# import Neovim: get_cursor, set_cursor, command, input
```
Comment (smolck) [inline src/repl.jl:13]: Can remove these comments

File: src/repl.jl [hunk 7]
```diff
@@ -3,27 +3,34 @@
 # Then, at any time, press Ctrl-O to edit in normal mode.
 # To be able to execute code and search history etc, going back to insert mode is neccesary.
 # For the moment, the first keypress after going to insert mode disappears sometimes.
-import Base: LineEdit, REPL
-using Neovim
-import Neovim: get_cursor, set_cursor, command, input
+# import Base: LineEdit, REPL
```
Comment (smolck) [inline src/repl.jl:6]: No need to keep comment

File: src/plugin_host.jl [hunk 8]
```diff
@@ -119,7 +120,7 @@ macro fnsync(args...)
 end
 
 function fun(ex)
-    if ex.head == :block && length(ex.args) == 2 && ex.args[1].head == :line
+    if ex.head == :block && length(ex.args) == 2 # && ex.args[2].head == :line
```
Comment (smolck) [inline src/plugin_host.jl:123]: Remove comment

File: src/plugin_host.jl [hunk 9]
```diff
@@ -1,18 +1,19 @@
-start_host() = try wait(nvim_child(HostHandler())) end
+# TODO(smolck): Didn't have `catch;`, is this right?
+start_host() = try wait(nvim_child(HostHandler())) catch; end
```
Comment (smolck) [inline src/plugin_host.jl:2]: Should probably handle this error properly?

File: test/child.vim [hunk 10]
```diff
@@ -1,5 +1,5 @@
 " FIXME: resolve path to script
-let g:julia_channel = rpcstart("julia", ["../test/child.jl"])
+let g:julia_channel = jobstart(["julia", "../test/child.jl"], {'rpc': v:true})
```
Comment (smolck) [inline test/child.vim:2]: Is this FIXME still relevant?

File: src/Neovim.jl [hunk 11]
```diff
@@ -9,29 +9,32 @@ export Buffer, Tabpage, Window
...
 end
 
 nvim_env(args...) = nvim_connect(ENV["NVIM_LISTEN_ADDRESS"], args...)
 
 function nvim_spawn(args...; cmd=`nvim --embed`)
-    output, input, proc = readandwrite(cmd)
-    (NvimClient(input, output, args...), proc)
+    # TODO(smolck): Just use the Process handle instead of p.in and p.out
+    #output, input, proc = readandwrite(cmd)
+    # (NvimClient(p, args...), proc)
```
Comment (smolck) [inline src/Neovim.jl:30]: Can remove comments

File: src/api_gen.jl [hunk 12]
```diff
@@ -1,27 +1,34 @@
...
-MsgPack.pack{N}(s, o::NvimApiObject{N}) = MsgPack.pack(s, Ext(N, o.hnd))
+# Not really module-interface clean, I know...
+# TODO(smolck): Was just s with Any type, is this right?
+MsgPack.pack(s::IO, o::NvimApiObject{N}) where {N} = MsgPack.pack(s, Extension(N, o.hnd))
 
-symbolize(val::Dict) = Dict{Symbol,Any}([(symbolize(k),symbolize(v)) for (k,v) in val])
-symbolize(val::Vector{UInt8}) = symbol(bytestring(val))
-symbolize(val::ByteString) = symbol(val)
+symbolize(val::Dict) = Dict{Symbol,Any}([(symbolize(k), symbolize(v)) for (k, v) in val])
+# symbolize(val::Vector{UInt8}) = symbol(bytestring(val))
```
Comment (smolck) [inline src/api_gen.jl:23]: Need to remove comment. And likewise for several others in this file.

File: src/client.jl [hunk 13]
```diff
@@ -2,22 +2,25 @@ const REQUEST = 0
 const RESPONSE = 1
 const NOTIFICATION = 2
 
-type NvimClient{S} <: NvimObject
-    input::S #input to nvim
+import Distributed.RemoteChannel
+
+mutable struct NvimClient{S} <: NvimObject
```
Comment (smolck) [inline src/client.jl:7]: Think I’ll leave removing the generic for another PR

**Review by smolck — COMMENTED** [review]
Didn’t go through everything in detail, but these are just things I noticed while scrolling through.

**Comment by bfredl** [comment]
I'm just gonna merge this, I trust @smolck to open follow up PR:s for remaining issues :]

**Comment by smolck** [comment]
@bfredl Sounds good, I'll also see if I can get the package in the packages repo now that it's 1.0 compatible (#10).

File: src/repl.jl (hunk 6 above)
Comment (smolck) [reply src/repl.jl:13]: Fixed in https://github.com/bfredl/Neovim.jl/commit/97ae01b61b6696dcbf3ed8ccfd285658311cb256

File: src/repl.jl (hunk 7 above)
Comment (smolck) [reply src/repl.jl:6]: Fixed in https://github.com/bfredl/Neovim.jl/commit/97ae01b61b6696dcbf3ed8ccfd285658311cb256

File: src/plugin_host.jl (hunk 8 above)
Comment (smolck) [reply src/plugin_host.jl:123]: Fixed in https://github.com/bfredl/Neovim.jl/commit/97ae01b61b6696dcbf3ed8ccfd285658311cb256

File: src/Neovim.jl (hunk 11 above)
Comment (smolck) [reply src/Neovim.jl:30]: Fixed in https://github.com/bfredl/Neovim.jl/commit/97ae01b61b6696dcbf3ed8ccfd285658311cb256

File: src/api_gen.jl (hunk 12 above)
Comment (smolck) [reply src/api_gen.jl:23]: Fixed in https://github.com/bfredl/Neovim.jl/commit/97ae01b61b6696dcbf3ed8ccfd285658311cb256
